File: pkg/sdk/context.go
package sdk

import (
	"context"
	"os"
)

type contextKey struct{}

// DefaultHeaderKey is used to extract the preview environment name from the X-Diverge-Env HTTP header.
const DefaultHeaderKey = "x-diverge-env"

// WithEnvironment returns a context with the environment name set.
func WithEnvironment(ctx context.Context, envName string) context.Context {
	return context.WithValue(ctx, contextKey{}, envName)
}

// EnvironmentFromContext extracts the environment name from the context.
func EnvironmentFromContext(ctx context.Context) string {
	if val, ok := ctx.Value(contextKey{}).(string); ok {
		return val
	}
	return ""
}

// GetHeaderKey returns the header key used for environment propagation.
// It defaults to "x-preview-env" if DIVERGE_HEADER_KEY is not set.
func GetHeaderKey() string {
	if val := os.Getenv("DIVERGE_HEADER_KEY"); val != "" {
		return val
	}
	return "x-preview-env"
}

File: pkg/sdk/node/src/context.ts
import { AsyncLocalStorage } from 'node:async_hooks';

export const DEFAULT_HEADER_KEY = 'x-diverge-env';
export const BINARY_HEADER_KEY = 'x-diverge-context-bin';

const storage = new AsyncLocalStorage<string>();

export function getEnvironment(): string {
  const env = storage.getStore();
  if (env) {
    return env;
  }
  return process.env.DIVERGE_ENV || "";
}

export function setEnvironment<T>(name: string, fn: () => T): T {
  return storage.run(name, fn);
}

export function withEnvironment(name: string): void {
  storage.enterWith(name);
}

export function getHeaderKey(): string {
  return process.env.DIVERGE_HEADER_KEY || DEFAULT_HEADER_KEY;
}
